kernel/kernel-api/src/main/java/io/delta/kernel/internal/IcebergCompatV2Utils.java

```diff
@@ -33,4 +66,216 @@ public static void validDataFileStatus(DataFileStatus dataFileStatus) {
+              // TODO: assign column mapping field ids
+              return Optional.empty();
+            }
+          });
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_NO_COMPAT_V1_ENABLED =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          if (Boolean.valueOf(
+              newMetadata
+                  .getConfiguration()
+                  .getOrDefault("delta.enableIcebergCompatV1", "false"))) {
+            throw new IllegalArgumentException(
+                "Not expected to have delta.enableIcebergCompatV1 enabled");
+          }
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_SUPPORTED_TYPES =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_ALLOWED_PARTITION_TYPES =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_NO_PARTITION_EVOLUTION =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static final IcebergCompatCheck ICEBERG_COMPAT_V2_CHECK_HAS_NO_DELETION_VECTORS =
+      new IcebergCompatCheck() {
+        @Override
+        void check(boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+          // TODO
+        }
+      };
+
+  private static class IcebergCompatMetadataValidatorAndUpdater {
+    private final TableConfig<Boolean> requiredDeltaTableProperty;
+    private final List<TableFeature> requiredDependencyTableFeatures;
+    private final List<RequiredDeltaTablePropertyEnforcer> requiredDeltaTableProperties;
+    private final List<IcebergCompatCheck> icebergCompatChecks;
+
+    IcebergCompatMetadataValidatorAndUpdater(
+        TableConfig<Boolean> requiredDeltaTableProperty,
+        List<TableFeature> requiredDependencyTableFeatures,
+        List<RequiredDeltaTablePropertyEnforcer> requiredDeltaTableProperties,
+        List<IcebergCompatCheck> icebergCompatChecks) {
+      this.requiredDeltaTableProperty = requiredDeltaTableProperty;
+      this.requiredDependencyTableFeatures = requiredDependencyTableFeatures;
+      this.requiredDeltaTableProperties = requiredDeltaTableProperties;
+      this.icebergCompatChecks = icebergCompatChecks;
+    }
+
+    boolean isEnabled(Metadata metadata) {
+      return requiredDeltaTableProperty.fromMetadata(metadata);
+    }
+
+    Optional<Metadata> validateAndUpdateMetadata(
```

Comment:
Do we also need to update the protocol here like in delta-spark or will we do that elsewhere? (to enable dependent features)

Reply:
Agreed. I like only updating the protocol in one place!